src/main/java/fi/aalto/cs/apluscourses/ui/ideactivities/OverlayPane.java

```diff
@@ -0,0 +1,201 @@
+  protected void paintComponent(Graphics graphics) {
+    super.paintComponent(graphics);
+
+    for (var c : popups) {
+      c.recalculateBounds();
+    }
+
+    // use Graphics's copy so as not to influence subsequent calls using the same Graphics object
+    Graphics2D g = (Graphics2D) graphics.create();
+    g.setComposite(AlphaComposite.SrcOver.derive(0.7f));
+    g.setColor(Color.BLACK); // using JBColor.BLACK is wrong here
+
+    var overlayArea = new Area(new Rectangle(0, 0, getWidth(), getHeight()));
+
+    for (var c : exemptComponents) {
+      // convertPoint is necessary because the component uses a different coordinate origin
+      var windowPos = SwingUtilities.convertPoint(c, c.getX(), c.getY(), this);
+      var componentRect = new Rectangle(windowPos.x, windowPos.y, c.getWidth(), c.getHeight());
+
+      overlayArea.subtract(new Area(componentRect));
+    }
+
+    for (var c : popups) {
+      // popups are already places in the overlay's coordinate system
+      var componentRect = new Rectangle(c.getX(), c.getY(), c.getWidth(), c.getHeight());
+      overlayArea.subtract(new Area(componentRect));
+    }
+
+    g.fill(overlayArea);
+  }
+
+  @Override
+  public JRootPane getRootPane() {
+    return associatedRootPane;
+  }
+
+  @Override
+  public int getWidth() {
+    return associatedRootPane.getLayeredPane().getWidth();
+  }
+
+  @Override
+  public int getHeight() {
+    return associatedRootPane.getLayeredPane().getHeight();
+  }
+
+  @Override
+  public Dimension getSize() {
+    return new Dimension(getWidth(), getHeight());
+  }
+
+  @Override
+  public Dimension getPreferredSize() {
+    return getSize();
+  }
+
+  @Override
+  public Dimension getMinimumSize() {
+    return getSize();
+  }
+
+  @Override
+  public Dimension getMaximumSize() {
+    return getSize();
+  }
+
+  @Override
+  public boolean isOpaque() {
+    return false;
+  }
+
+  @Override
+  public boolean isVisible() {
+    return true;
+  }
+
+  public static boolean isOverlayInstalled() {
+    return activeOverlay != null;
+  }
+
+  /**
+   * Installs the overlay.
+   */
+  public static void installOverlay() {
+    if (isOverlayInstalled()) {
+      throw new IllegalStateException("An overlay is already installed");
+    }
+
+    activeOverlay = new OverlayPane();
+    activeOverlay.getRootPane().getLayeredPane().add(activeOverlay, 20000);
+    activeOverlay.revalidatePane();
+  }
+
+  /**
+   * Removes the overlay.
+   */
+  public static void removeOverlay() {
+    if (!isOverlayInstalled()) {
+      throw new IllegalStateException("No overlay is currently installed");
+    }
+
+    activeOverlay.getRootPane().getLayeredPane().remove(activeOverlay);
+    for (var c : activeOverlay.popups) {
+      activeOverlay.getRootPane().getLayeredPane().remove(c);
+    }
+    activeOverlay.revalidatePane();
+
+    activeOverlay = null;
+  }
+
+  /**
+   * Marks a component not to be dimmed.
+   */
+  public static void showComponent(Component c) {
```

Comment:
Nit: The overlay could be accessed through an object instance instead of static methods.